Inline Help: extract Popover content to separate component

jsnajdr opened the pull request:
This is one of the crucial parts of #55189: extracting the Inline Help popover content from the actual `Popover` component. This content component will be rendered, without the parent `Popover` component, in the Inline Help widget.

**How to test and review:**
Verify that all needed props are passed between the `Popover` and `Content` components. Verify that the split of the SCSS stylesheet into two is done correctly. In `style.scss`, there remain only styles for the round question mark button and for the popover positioning, and everything about the content is moved the the `popover-content.scss` module.

There are two commits that remove unused CSS styles: for switching between Classic and Block editor, and for a notification dot on the help button that used to be used when Site Checklist was integrated with the Inline Help widget.

jsnajdr committed:
Inline Help: remove unused styles for Classic/Block button

jsnajdr committed:
Inline Help: remove unused notification style

jsnajdr committed:
InlineHelp: extract popover content to separate component (step 1)

jsnajdr committed:
InlineHelp: extract popover content to separate component (step 2)

jsnajdr commented:
There are e2e failures when clicking on the "Show me where..." search results which is weird because they're working for me when I'm clicking on them as a human. Investigating...

jsnajdr committed:
Streamline the InlineHelpCompactResults component

jsnajdr committed:
Distribute InlineHelp search styles to individual components

jsnajdr committed:
Remove the unused .inline-help__search .button style: there's no button

jsnajdr committed:
Increase specificity of .search-card styles to avoid order ambiguity

jsnajdr commented:
So, the e2e tests were failing not in the Inline Help popover itself, but on the Home page when doing a help search there:

<img width="615" alt="Screenshot 2022-01-03 at 12 28 49" src="https://user-images.githubusercontent.com/664258/147926500-946e79b5-980a-40ea-b637-dbad2e76ea02.png">

And the reason was that the PR moved some of the `.inline-help__results` and `inline-help__search` styles to the `popover-content.scss` file, and these styles were not loaded on the Home page. They started to be async-loaded only after actually opening the popover.

And the e2e test depends on some of these styles being applied and active. I didn't look into the exact reasons, but it's most likely something with checking whether an element is visible or not, which is a CSS-influenced thing.

The solution is to move some styles from `popover-content.scss` to component-specific stylesheets that are imported by the affected component itself.

After this fix all checks are green again and the PR is ready for review :+1:

tyxla commented on client/blocks/inline-help/inline-help-compact-results.jsx:
```diff
@@ -1,30 +1,29 @@
+	return (
+		<ul className="inline-help__results-list">
+			{ helpLinks.map( ( link ) => (
+				<li key={ link.link + '#' + link.id } className="inline-help__results-item">
```
It's not ideal that those live in 2 places. Ideally, they should be managed by a single component, and whether we'll have a "inline-help__results-cell" wrapper (and the other differences) would be determined by props.

tyxla commented on client/blocks/inline-help/inline-help-search-card.jsx:
```diff
@@ -6,6 +6,7 @@ import { useDispatch } from 'react-redux';
 import SearchCard from 'calypso/components/search-card';
 import { useHelpSearchQuery } from 'calypso/data/help/use-help-search-query';
 import { recordTracksEvent } from 'calypso/state/analytics/actions';
+import './inline-help-search-card.scss';
```
The `inline-help__search` classname under which all CSS inside this stylesheet is nested is not being used directly in this component, so at first glance it feels odd to import this stylesheet here.

jsnajdr replied:
Yes, it's used as:
```jsx
<div className="inline-help__search">
  <HelpSearchCard />
  <HelpSearchResults />
</div>
```
It doesn't have its own component where the styles could live, but always travels together with the `HelpSearchCard` component. So including the styles there is a dirty trick, but it works. Can be cleanup as one of follow-ups.

tyxla commented on client/blocks/inline-help/inline-help-search-results.scss:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	a {
+		//color: var( --color-primary );
```
I think we can get rid of this one.

jsnajdr replied:
Removed the commented-out styles in 15c1a008e8

tyxla commented on client/blocks/inline-help/inline-help-search-results.scss:
```diff
@@ -0,0 +1,91 @@
+		display: flex;
+		padding: 8px 16px;
+
+		span {
```
It's a bummer we're using spans and not a classnames for the title wrapper. Pre-existing, so feel free to ignore.

tyxla commented on client/blocks/inline-help/popover-content.scss:
```diff
@@ -0,0 +1,168 @@
+	text-align: left;
+}
+
+.inline-help__contact {
```
At first glance, it appears that this classname is not used anywhere, but at a second glance it appears to be used here:

https://github.com/Automattic/wp-calypso/blob/a82f8469a286519ffe3e68793b5d61778cf34e07/client/blocks/inline-help/popover-content.jsx#L147

tyxla commented on client/blocks/inline-help/style.scss:
```diff
@@ -6,12 +6,6 @@
 }
 
-.button.inline-help__classic-editor-toggle,
-.button.inline-help__gutenberg-editor-toggle {
```
Confirmed to be unused 👍

tyxla commented on client/blocks/inline-help/style.scss:
```diff
@@ -68,19 +62,6 @@
 		border-color: var( --color-primary-40 );
 	}
-
-	&.has-notification::before {
```
Confirmed to be unused. Seems like we can also remove it from here:

https://github.com/Automattic/wp-calypso/blob/7fcaec0432d9d76c548f2fe998025d691a7f79ea/client/assets/stylesheets/directly.scss#L160

jsnajdr replied:
That seems to be a very different place, inside the Directly widget's markup. There's a `.thread-view-title-container.has-notification` class, displaying notifications on a discussion thread?

tyxla commented on client/blocks/inline-help/popover-content.scss:
```diff
@@ -0,0 +1,168 @@
+.inline-help__popover-content {
+	background-color: var( --color-surface );
+
+	p {
```
Noting that these `p` and `a, button` styles were previously applied to `.inline-help__mobile-overlay`. Should we preserve that?

jsnajdr replied:
No, they were applied to `.inline-help__popover.popover`. But the GitHub diff viewer collapses the unchanged parts of `style.scss` so that it appears they were inside `.inline-help__mobile-overlay`. Please have one more look.

tyxla commented on client/blocks/inline-help/popover-content.scss:
```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+.inline-help__view-heading {
```
This lives only in the `InlineHelpForumView` component. Should we move it to its own stylesheet there?

jsnajdr replied:
Moved in e48c4fbe49

tyxla commented on client/blocks/inline-help/popover-content.scss:
```diff
@@ -0,0 +1,168 @@
+	margin-bottom: 5px;
+}
+
+.inline-help__richresult {
```
This one doesn't seem to be used right now. Actually, it seems like it was never used since it was introduced in #24390.

jsnajdr replied:
This is used, just like the `inline-help__contact` class. It's one of the secondary views styled as `inline-help__${ this.state.activeSecondaryView }`

tyxla commented on client/my-sites/customer-home/cards/features/help-search/style.scss:
```diff
@@ -159,7 +159,7 @@ $min_results_height: 180px;
 	}
 
-	.search-card {
+	.card.search-card {
```
We'll need to test all affected flows thoroughly. Stylesheet load order is known to be different between development and production, so there might be subtle unexpected differences, caused by different load order of stylesheets.

jsnajdr replied:
I added these `.card` classes when I noticed that the search results styles are sometimes broken. See the padding between the search input and the results on the screenshot in this comment: https://github.com/Automattic/wp-calypso/pull/59413#issuecomment-1004037094 The padding is too big, due to style order ambiguity. Agreed that it needs more testing.

tyxla reviewed (commented):
Thanks for splitting the giant stylesheet, this is great work!

I've left some comments, observations and suggestions after taking a more thorough look.

Also, I'd really appreciate it if we can have thorough testing instructions in the PR, because we'll need them, especially when merging, in order to ensure that stylesheet loading order doesn't create problems in production.

jsnajdr committed:
Inline the richresult title className

jsnajdr committed:
Move forum view styles to its own module

jsnajdr committed:
Remove commented-out styles

jsnajdr commented:
**Testing instruction:**
Verify all the places where help search and results are displayed. There are plenty of them!

In the Inline Help popover, there's the default view with help search:

<img width="326" alt="Screenshot 2022-01-05 at 13 48 59" src="https://user-images.githubusercontent.com/664258/148220330-0e772185-5934-4887-939a-b5b54fdef98a.png">

After clicking "Contact us" and typing a message in the chat box, you'll see some contextual results before your message is really sent:

<img width="328" alt="Screenshot 2022-01-05 at 13 48 25" src="https://user-images.githubusercontent.com/664258/148220421-36237ad1-2521-4b2e-abf7-3b2c9be4c22d.png">

Then there are the `/help` pages. On the `/help` page, there is help search, which shows some suggestions (rendered using the "help results" components) even before you start searching:

<img width="760" alt="Screenshot 2022-01-05 at 13 51 48" src="https://user-images.githubusercontent.com/664258/148220627-b59606a3-04d5-4653-9d9b-6a55b2ed3f9e.png">

After you start searching, some "real" results are shown:

<img width="754" alt="Screenshot 2022-01-05 at 13 53 16" src="https://user-images.githubusercontent.com/664258/148220751-c56498e1-2826-4e42-b030-381c9ee94c1d.png">

Then after clicking "Contact support" you'll be taken to `/help/contact`, and while typing into the chat message box, you'll see some contextual results again:

<img width="721" alt="Screenshot 2022-01-05 at 13 55 54" src="https://user-images.githubusercontent.com/664258/148221131-88b314bf-6341-4e84-a471-f8e3ea8b3c07.png">

Finally there's the Home screen where you can also search help:

<img width="477" alt="Screenshot 2022-01-05 at 13 57 24" src="https://user-images.githubusercontent.com/664258/148221247-fd67ca03-7412-43c0-85cf-01840ee6edc0.png">

All these places share some common styles (like `.help-search .card.search-card`) and can be affected when the styles are modified.

jsnajdr commented:
Howdy @tyxla :wave: I addressed all the review comments, and also posted some detailed testing instructions. I believe the PR is ready for another round of review.

tyxla reviewed (approved):
Code looks good! 

I've also done some thorough testing and it works great!

It could use a rebase before landing (I noticed that the bug you fixed in the footer in #59668 was still not in this branch).

Nice work 💯

jsnajdr commented:
Rebased this onto the latest `trunk` (the base commit was indeed quite old) and will merge once all checks come back green :+1:
